Add PolicyType and Policy models

m4dcoder opened the pull request:
Add appropriate DB, persistence, and API models, REST controllers, and CLI operations for PolicyType and Policy.

manasdk commented on st2api/st2api/controllers/v1/policies.py:
```diff
@@ -0,0 +1,263 @@
+        """
+        op = 'POST /policies/'
+
+        if not hasattr(instance, 'enabled'):
```
Could be handled in DB?

m4dcoder replied:
Yes.  I just copied this from https://github.com/StackStorm/st2/blob/master/st2api/st2api/controllers/v1/actions.py#L80.

manasdk commented on st2api/st2api/controllers/v1/policies.py:
```diff
@@ -0,0 +1,263 @@
+        db_model_id = db_model.id
+
+        try:
+            validate_not_part_of_system_pack(db_model)
```
Why not?

m4dcoder replied:
I copied this from the ContentPack controller.  I guess I can use cases where users may want to either delete and change the policies that are deployed as part of the core pack.  But it is going to be inconsistent with how resources (i.e. actions) are managed under the core pack.  Are we ok with that?

manasdk replied:
Policies are fundamentally deployment specific. Heck even the fact that `system` actions cannot be changed is broken in my view.

m4dcoder replied:
Another question, what should the behavior be for a st2 upgrade?  If users modified/deleted the policies in core, during an upgrade, should we overlay an existing installation?

manasdk commented on st2api/st2api/controllers/v1/policies.py:
```diff
@@ -0,0 +1,263 @@
+    access = Policy
+
+    supported_filters = {
+        'pack': 'pack'
```
`resource_ref ` could also be a useful filter

m4dcoder replied:
Good point!

manasdk commented on st2client/st2client/client.py:
```diff
@@ -73,55 +73,79 @@ def __init__(self, base_url=None, auth_url=None, api_url=None, api_version=None,
-            models.Token, self.endpoints['auth'], cacert=self.cacert, debug=self.debug)
-        self.managers['RunnerType'] = models.ResourceManager(
-            models.RunnerType, self.endpoints['api'], cacert=self.cacert, debug=self.debug)
+
```
Would be nice to avoid stylistic changes if the affected files > 5.

m4dcoder replied:
Ok.  I'll keep that in mind.  It is too tempting.  The stuffs are not sorted and it was hard to read.... :P

manasdk replied:
I understand the temptation but if you believe that you would like the code to be reviewed then keeping in mind that this adds noise and make reviewing harder is also important.

m4dcoder replied:
Got it.

lakshmi-kannan replied:
Plus you are gonna see merge conflict with https://github.com/StackStorm/st2/pull/1480 :P.

m4dcoder replied:
Hm, maybe I should merge now. :p

manasdk commented on st2common/st2common/models/db/policy.py:
```diff
@@ -0,0 +1,180 @@
+
+    def __init__(self, *args, **kwargs):
+        super(PolicyTypeDB, self).__init__(*args, **kwargs)
+        self.ref = PolicyTypeReference.to_string_reference(resource_type=self.resource_type,
```
The fact that this ref is diffrent from other ref in the system bothers me. This ref does not follows the standard `pack.name` convention either.

I see the value in scope by name of the type but couldn't that be simply captured in the name? To deviate from the usual form of ref which is basically the way to uniquely identify, in a human readable form, objects in our system we need absolute conviction. I am not convinced.

m4dcoder replied:
I thought about that and started initially to name the policy like "action.concurrency".  But how is it going to be enforced?  Just checking the name and then throw an exception that user needs to prefix it with "action" (or other type)?

manasdk replied:
Why enforce?

m4dcoder replied:
> I see the value in scope by name of the type
I prefer organization and cleanliness.  Enforcing some naming standard makes the UI cleaner.  But if this is causing a lot of heartache, I can remove it.

manasdk replied:
> organization and cleanliness

organization and cleanliness are subjective for the most part. Carrot and no stick would be my approach. Same property different meaning depending on the object leads to confusion. 

Anyway we are discussing 2 separate matters -
1. ref property meaning something different PolicyType. Should not.
2. Enforcing a specific structure for any naming property - that is arguable but I am not against.

manasdk commented:
I am +1 to the overall change.

I am not in favor of `PolicyTypeReference` as meaning anything other than ref i.e. `pack.name`.

lakshmi-kannan commented on st2api/st2api/controllers/resource.py:
```diff
@@ -188,7 +188,7 @@ def _get_by_name(self, resource_name):
 
 
-class ContentPackResourceControler(ResourceController):
+class ContentPackResourceController(ResourceController):
```
These changes should definitely be in its own PR.

lakshmi-kannan commented on st2api/st2api/controllers/v1/policies.py:
```diff
@@ -0,0 +1,263 @@
+        return self.model.from_model(db_model)
+
+    @jsexpose(arg_types=[str], status_code=http_client.NO_CONTENT)
+    def delete(self, ref_or_id):
```
I am not sure if we should support delete this early. Don't we have to delete all references to the policy before deleting the policy itself?

m4dcoder replied:
Do you mean the other way around?  Policy overlays on top of a resource (i.e. action).  Deleting a policy should be independent from the action.  On the other hand, if a user deletes an action for example, should all the existing policies that reference the action be deleted as well?  At this point (experimental), I don't think it matters.  If the action is deleted, the action would never be executed, and any policies that reference the action won't be touched.

lakshmi-kannan replied:
I guess I don't how the association is going to established between policies and content at this point. So yeah, deleting a policy cannot be simple as just deleting the model from db. It probably has more logic to it is all I am saying. I am fine with leaving the code as such.

m4dcoder replied:
Sure. When we can articulate what that association is, we can add logic to the delete.

manasdk commented:
I guess we all have decided not to block on PolicyTypeReference.

m4dcoder commented:
Except for validate_not_part_of_system_pack(db_model), I addressed comments above. If we want to allow PUT/DELETE for policies in system pack, I can remove validate_not_part_of_system_pack(db_model) in a separate commit.
